File: Algorithm/Python/1540.can-convert-string-in-k-moves.py
#
# @lc app=leetcode id=1540 lang=python
#
# [1540] Can Convert String in K Moves
#

# @lc code=start

# use fixed-len counter instead of set to avoid TLE
class Solution(object):
    def canConvertString(self, s, t, k):
        """
        :type s: str
        :type t: str
        :type k: int
        :rtype: bool
        """
        if len(s) != len(t):
            return False
        
        counter = [0] * 26
        for i in range(len(s)):
            diff = (ord(t[i]) - ord(s[i]) + 26) % 26
            if diff > 0 and diff + counter[diff] * 26 > k:
                return False
            counter[diff] += 1
        return True


# # TLE
# class Solution(object):
#     def canConvertString(self, s, t, k):
#         """
#         :type s: str
#         :type t: str
#         :type k: int
#         :rtype: bool
#         """
#         if len(s) != len(t):
#             return False
#         used = set()
#         for i in range(len(s)):
#             diff = ord(t[i]) - ord(s[i])
#             # print("{}:{}".format(s[i], diff))
#             if diff != 0:
#                 convert = False
#                 while diff <= k:
#                     if diff > 0 and diff not in used:
#                         used.add(diff)
#                         convert = True
#                         break
#                     diff += 26
#                 if not convert:
#                     return False                    
#         return True
     
# @lc code=end

File: Algorithm/C++/1540.can-convert-string-in-k-moves.cpp
/*
 * @lc app=leetcode id=1540 lang=cpp
 *
 * [1540] Can Convert String in K Moves
 */

// @lc code=start

#include <string>
#include <vector>
#include <unordered_set>


// use fixed-len counter instead of set to avoid TLE
class Solution {
public:
    bool canConvertString(std::string s, std::string t, int k) {
        if (s.size() != t.size()) {
            return false;
        }
        std::vector<int> counter(26, 0);
        int diff;
        for (int i = 0; i < s.size(); i++) {
            diff = (t[i] - s[i] + 26) % 26; // avoid negative
            if (diff > 0 && diff + counter[diff] * 26 > k) {
                return false;
            }
            counter[diff]++;
        }
        return true;
    }
};

/* use set will TLE
class Solution {
public:
    bool canConvertString(std::string s, std::string t, int k) {
        if (s.size() != t.size()) {
            return false;
        }
        
        std::unordered_set<int> used;
        int diff = 0;
        bool convert;
        for (int i = 0; i < s.size(); i++) {
            convert = false;
            diff = t[i] - s[i];
            if (diff != 0) {
                while (diff <= k) {
                    if (diff  > 0 && used.find(diff) == used.end()) {
                        used.insert(diff);
                        convert = true;
                        break;
                    }
                    diff += 26;
                }
                if (!convert) {
                    return false;
                }
            } 
        }
        return true;
    }
};
*/

// @lc code=end

